go/chat/push.go

```diff
@@ -6,292 +6,483 @@ import (
 		return err
 	}
-	if len(inbox.Convs) != 1 {
-		g.Debug(ctx, "resolve: unable to find conversation")
-		return fmt.Errorf("unable to find conversation")
-	}
-	updateConv := inbox.Convs[0]
+	uid := gregor1.UID(m.UID().Bytes())
+
+	// Order updates based on inbox version of the update from the server
+	cb := g.orderer.WaitForTurn(ctx, uid, update.InboxVers)
+	bctx := BackgroundContext(ctx, g.G().GetEnv())
+	go func(ctx context.Context) {
+		defer g.Trace(ctx, func() error { return nil }, "TlfResolve(goroutine)")()
+		<-cb
+		g.Lock()
+		defer g.Unlock()
+		defer g.orderer.CompleteTurn(ctx, uid, update.InboxVers)
+		// Get and localize the conversation to get the new tlfname.
+		inbox, _, err := g.G().InboxSource.Read(ctx, uid, nil, true, &chat1.GetInboxLocalQuery{
+			ConvIDs: []chat1.ConversationID{update.ConvID},
+		}, nil)
+		if err != nil {
+			g.Debug(ctx, "resolve: unable to read conversation: %s", err.Error())
+			return
+		}
+		if len(inbox.Convs) != 1 {
+			g.Debug(ctx, "resolve: unable to find conversation")
+			return
+		}
+		updateConv := inbox.Convs[0]
 
-	resolveInfo := chat1.ConversationResolveInfo{
-		NewTLFName: updateConv.Info.TlfName,
-	}
+		resolveInfo := chat1.ConversationResolveInfo{
+			NewTLFName: updateConv.Info.TlfName,
+		}
 
-	g.G().NotifyRouter.HandleChatTLFResolve(context.Background(), keybase1.UID(uid), update.ConvID, resolveInfo)
+		g.G().NotifyRouter.HandleChatTLFResolve(ctx, keybase1.UID(uid.String()),
+			update.ConvID, resolveInfo)
+	}(bctx)
 
 	return nil
 }
 
 func (g *PushHandler) Activity(ctx context.Context, m gregor.OutOfBandMessage, badger *badges.Badger) (err error) {
+	var identBreaks []keybase1.TLFIdentifyFailure
+	ctx = Context(ctx, g.G().GetEnv(), keybase1.TLFIdentifyBehavior_CHAT_GUI, &identBreaks,
+		g.identNotifier)
 	defer g.Trace(ctx, func() error { return err }, "Activity")()
 	if m.Body() == nil {
 		return errors.New("gregor handler for chat.activity: nil message body")
 	}
 
+	// Decode into generic form
 	var activity chat1.ChatActivity
 	var gm chat1.GenericPayload
+	uid := m.UID().Bytes()
 	reader := bytes.NewReader(m.Body().Bytes())
 	dec := codec.NewDecoder(reader, &codec.MsgpackHandle{WriteExt: true})
 	err = dec.Decode(&gm)
 	if err != nil {
+		g.Debug(ctx, "chat activity: failed to decode into generic payload: %s", err.Error())
 		return err
 	}
+	g.Debug(ctx, "chat activity: action %s vers: %d", gm.Action, gm.InboxVers)
+
+	// Order updates based on inbox version of the update from the server
+	cb := g.orderer.WaitForTurn(ctx, uid, gm.InboxVers)
+	bctx := BackgroundContext(ctx, g.G().GetEnv())
+	go func(ctx context.Context) {
+		defer g.Trace(ctx, func() error { return nil }, "Activity(goroutine)")()
+		<-cb
+		g.Lock()
+		defer g.Unlock()
+		defer g.orderer.CompleteTurn(ctx, uid, gm.InboxVers)
+
+		action := gm.Action
+		reader.Reset(m.Body().Bytes())
+		switch action {
+		case "newMessage":
+			var nm chat1.NewMessagePayload
+			err = dec.Decode(&nm)
```

Comment:
Yeah good catch, admittedly I kind of dropped all that code that was there into a goroutine. I'll do a second pass to make sure there isn't anything else like this left over.